src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClientX.Handlers.Connection;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    internal class TransportCreationHandler : IHandler<ConnectionRequest>
+    {
+        /// <inheritdoc />
+        public IHandler<ConnectionRequest> NextHandler { get; set; }
+
+        /// <inheritdoc />
+        public async ValueTask Handle(ConnectionRequest request, bool isAsync, CancellationToken ct)
+        {
+            switch (request.TransportCreationParams)
+            {
+                case SharedMemoryTransportCreationParams:
+                    throw new NotImplementedException();
+
+                case NamedPipeTransportCreationParams:
+                    throw new NotImplementedException();
+
+                case TcpTransportCreationParams tcpParams:
+                    request.ConnectionStream = await HandleTcpRequest(tcpParams, ct).ConfigureAwait(false);
```

Comment:
I spent sometime reading up on this and digging through the code for the calls, but I'm not sure I see the sync-over-asyn anti-pattern going on. It looks like `Dns.GetHostAddressesAsync` isn't really async, but` Socket.ConnectAsync` is definitely async. I could change the call to get the addresses to use the sync method, but I think we're justified in using async for connecting. It sounds like you might know more about this than I do, so please let me know if I'm missing something.

This might also be a time for using one of those `ValueTask`s I've heard so much about 😅

Reply:
I'm thinking of a case where we call the method with `isAsync` set to false. Because `isAsync` is never passed down to `HandleTcpRequest`, it's effectively always asynchronous. A caller who waits on Handle is wrapping a synchronous operation around async operations occurring on the thread pool (sync-over-async.) Your points around `Dns.GetHostAddressesAsync` and `Socket.ConnectAsync` are absolutely correct, but protect against thread-level blocking within a task (async-over-sync.)

There's a minor (but immediate) performance hit because we might now allocate Tasks. There's also a more subtle one which is a pain point for the current iteration of SqlClient, which is that if the thread pool is heavily loaded it could be some time until the thread pool has enough capacity to start the work. A caller wouldn't necessarily expect that if MultiSubnetFailover is disabled. In the worst case, both sync-over-async and async-over-sync can cause deadlocks.

Reply:
I think the Sync path would need a separate flow where the sync versions of DNS and sockets are used, rather than relying on the async overloads. Likely in this PR it would make sense to `throw` or add a TODO to say Sync is to be supported. 

For MSF in sync mode, sync over async will suffer from the problems @edwardneal  mentioned. However a potential path forward for sync mode is to open all sockets in non-blocking mode, and later check using Socket.Select to see which one is writeable for the timeout duration, and return that particular socket. 
This way the I/O is used for socket open parallelization, and we dont rely on sync over async.

Reply:
Either a `throw` or a TODO is fine by me, so long as the problem's raised explicitly rather than implicitly - getting the correct behaviour from various scenarios of MSF/TNIR/connection timeouts would probably be better handled separately.